source/usr/lib/mcvirt/cluster/cluster.py

```diff
@@ -15,437 +17,555 @@
+        # Ensure the user has privileges to manage the cluster
+        self._get_registered_object('auth').assert_permission(PERMISSIONS.MANAGE_CLUSTER)
 
-        # Connect to each of the nodes
-        if (self.mcvirt_instance.initialise_nodes):
-            self.connectNodes()
+        try:
+            config_json = base64.b64decode(node_connection_string)
+            node_config = json.loads(config_json)
+            assert 'username' in node_config and node_config['username']
+            assert 'password' in node_config and node_config['password']
+            assert 'ip_address' in node_config and node_config['ip_address']
+            assert 'hostname' in node_config and node_config['hostname']
+            assert 'ca_cert' in node_config and node_config['ca_cert']
+        except:
+            raise InvalidConnectionString('Connection string is invalid')
 
-    def addNodeRemote(self, remote_host, remote_ip_address, remote_public_key):
-        """Adds the machine to a remote cluster"""
         # Determine if node is already connected to cluster
-        if (self.checkNodeExists(remote_host)):
-            raise NodeAlreadyPresent('Node %s is already connected to the cluster' % remote_host)
-        local_public_key = self.getSshPublicKey()
-        self.addNodeConfiguration(remote_host, remote_ip_address, remote_public_key)
-        return local_public_key
-
-    def addNode(self, remote_host, remote_ip, password):
-        """Connects to a remote MCVirt machine, shares SSH keys and clusters the machines"""
-        from remote import Remote
-        from mcvirt.node.drbd import DRBD
-        # Ensure the user has privileges to manage the cluster
-        self.mcvirt_instance.getAuthObject().assertPermission(Auth.PERMISSIONS.MANAGE_CLUSTER)
+        if self.check_node_exists(node_config['hostname']):
+            raise NodeAlreadyPresent(
+                'Node %s is already connected to the cluster' % node_config['hostname'])
 
-        # Determine if node is already connected to cluster
-        if (self.checkNodeExists(remote_host)):
-            raise NodeAlreadyPresent('Node %s is already connected to the cluster' % remote_host)
+        # Create CA public key for machine
+        ssl_object = self._get_registered_object(
+            'certificate_generator_factory'
+        ).get_cert_generator(node_config['hostname'])
+        ssl_object.ca_pub_file = node_config['ca_cert']
 
         # Check remote machine, to ensure it can be synced without any
         # conflicts
-        remote = Remote(self, remote_host, remote_ip=remote_ip, password=password,
-                        save_hostkey=True)
+        remote = Connection(username=node_config['username'], password=node_config['password'],
+                            host=node_config['hostname'])
         try:
-            self.checkRemoteMachine(remote)
+            self.check_remote_machine(remote)
         except:
-            remote = None
             raise
```

Comment:
Is there any point in this ``try ... except`` anymore? Because if any exception is caught it is just re-raised again straight away without doing anything.